### PR TITLE
openstack V3 authentication

### PR DESCRIPTION
**Pull request by cmouse**
## Contributor Checklist:

* [x] I have updated the unit tests
* [x] I have created a file in the `master/buildbot/newsfragment` directory (and read the `README.txt` in that directory)
* [x] I have updated the appropriate documentation

### CODE REVIEW

File: master/docs/manual/cfg-workers-openstack.rst [hunk 1]
```diff
@@ -64,10 +64,17 @@ These are the same details set in either environment variables or passed as opti
...
+``os_user_domain``
+
+``os_project_domain``
+
 ``os_auth_url``
     The OpenStack authentication needed to create and delete instances.
     These are the same as the environment variables with uppercase names of the arguments.
 
+``auth_plugin``
+    Auth plugin to use with keystoneauth1. Use ``v3password`` for V3 authentication.
```
Comment (tardyp) [inline master/docs/manual/cfg-workers-openstack.rst]: use 'password' for v2 authentication.

Comment (seankelly) [inline master/docs/manual/cfg-workers-openstack.rst]: Please include the default value.

File: master/buildbot/worker/openstack.py [hunk 2]
```diff
@@ -99,12 +102,12 @@ def __init__(self, name, password,
...
-    def _constructClient(client_version, username, password, project_name,
-                         auth_url):
+    def _constructClient(client_version, username, user_domain, password, project_name, project_domain,
+                         auth_url, auth_plugin):
         """Return a novaclient from the given args."""
-        loader = loading.get_plugin_loader('password')
-        auth = loader.load_from_options(auth_url=auth_url, username=username,
-                                        password=password, project_name=project_name)
+        loader = loading.get_plugin_loader(auth_plugin)
+        auth = loader.load_from_options(auth_url=auth_url, username=username, user_domain_name=user_domain,
```
Comment (tardyp) [inline master/buildbot/worker/openstack.py]: I am not sure this will support the former method as well.

Comment (seankelly) [inline master/buildbot/worker/openstack.py]: Do you know when `user_domain_name` and `project_domain_name` were introduced?

File: master/docs/manual/cfg-workers-openstack.rst (hunk 1 above)
Comment (cmouse) [inline master/docs/manual/cfg-workers-openstack.rst]: 'password' is default value.

File: master/buildbot/worker/openstack.py [hunk 3]
```diff
@@ -58,7 +58,10 @@ def __init__(self, name, password,
                  os_username,
                  os_password,
                  os_tenant_name,
+                 os_user_domain=None,
```
Comment (seankelly) [inline master/buildbot/worker/openstack.py]: This is a syntax error. Parameters with default values must come after those without defaults.

File: master/buildbot/worker/openstack.py (hunk 2 above)
Comment (cmouse) [inline master/buildbot/worker/openstack.py]: I don't know other than that they have been part of V3 keystone API always.

Comment (cmouse) [inline master/buildbot/worker/openstack.py]: Have to find someone with v2 or v1 keystone to test this.

File: master/docs/manual/cfg-workers-openstack.rst (hunk 1 above)
Comment (cmouse) [inline master/docs/manual/cfg-workers-openstack.rst]: apparently it seems that using 'password' works for v3 as well.

**Comment by cmouse** [comment]
Closes #3008

File: master/buildbot/test/unit/test_worker_openstack.py [hunk 4]
```diff
@@ -71,6 +71,17 @@ def test_constructor_minimal(self):
...
+    def test_constructor_minimal_keystone_v3(self):
+        bs = openstack.OpenStackLatentWorker(
+            'bot', 'pass', os_user_domain='test', os_project_domain='test',
+            **self.bs_image_args)
+        self.assertEqual(bs.workername, 'bot')
+        self.assertEqual(bs.password, 'pass')
+        self.assertEqual(bs.flavor, 1)
+        self.assertEqual(bs.image, 'image-uuid')
+        self.assertEqual(bs.block_devices, None)
+        self.assertIsInstance(bs.novaclient, novaclient.Client)
```
Comment (seankelly) [inline master/buildbot/test/unit/test_worker_openstack.py]: There should be asserts for the two new parameters.

Comment (cmouse) [inline master/buildbot/test/unit/test_worker_openstack.py]: not sure how to assert them as they are not directly accessible.

Comment (seankelly) [inline master/buildbot/test/unit/test_worker_openstack.py]: The auth is saved in Session but that isn't saved in the client. Extend Client to save session so you can assert it.

**Comment by tardyp** [comment]
So where are we on this one? We do need to add two more asserts right?

**Comment by cmouse** [comment]
I suppose so, I'll do it soon.

File: master/buildbot/test/unit/test_worker_openstack.py (hunk 4 above)
Comment (cmouse) [reply master/buildbot/test/unit/test_worker_openstack.py]: assertions added

**Pull request merged**